File: frontend/src/schemas/decision.ts
import { z } from "zod";

export const decisionActionSchema = z.enum(["APPROVE", "RETURN", "ESCALATE"]);
export const decisionRequestSchema = z.object({
  action: decisionActionSchema,
  rationale: z.string().trim().min(1).max(2000),
  reason_code: z.string().regex(/^[A-Z][A-Z0-9_]{1,63}$/),
  proposal_id: z.string().uuid().nullable(),
}).strict();
export const decisionResponseSchema = z.object({
  decision_id: z.string().uuid(),
  exception_id: z.string().uuid(),
  run_id: z.string().uuid(),
  invoice_id: z.string().uuid(),
  action: decisionActionSchema,
  actor_id: z.string().min(1),
  stage: z.enum(["PENDING_SIGNOFF", "QUEUED_FOR_RESUME"]),
  proposal_id: z.string().uuid().nullable(),
}).strict();

export type DecisionRequest = z.infer<typeof decisionRequestSchema>;
export type DecisionResponse = z.infer<typeof decisionResponseSchema>;

File: src/invoiceops_agent/api/schemas/decision.py
"""Two-person exception decision transport contracts."""

from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator

DecisionAction = Literal["APPROVE", "RETURN", "ESCALATE"]
DecisionStage = Literal["PENDING_SIGNOFF", "QUEUED_FOR_RESUME"]


class DecisionRequest(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    action: DecisionAction
    rationale: str = Field(min_length=1, max_length=2000)
    reason_code: str = Field(pattern=r"^[A-Z][A-Z0-9_]{1,63}$")
    proposal_id: UUID | None = None

    @field_validator("rationale")
    @classmethod
    def nonblank_rationale(cls, value: str) -> str:
        cleaned = value.strip()
        if not cleaned:
            raise ValueError("Decision rationale must contain non-whitespace text")
        return cleaned


class DecisionResponse(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    decision_id: UUID
    exception_id: UUID
    run_id: UUID
    invoice_id: UUID
    action: DecisionAction
    actor_id: str
    stage: DecisionStage
    proposal_id: UUID | None
